### DISCUSSION TITLE
Is there an example of using HTTP/2?

### DISCUSSION THREAD

**Discussion by Nikaidou-Shinku** [tokio-rs/axum#1705] [Q&A] [answered]
I tried to enable the `http2` feature flag and used `http2_only(true)` but failed, my browser reported `ERR_INVALID_HTTP_RESPONSE`. Then I tried to disable the default features and enable the `http2` and `tokio`, but failed again.

I find nothing about HTTP/2 in the `examples` folder. Can someone give me an example?

**Comment by davidpdrsn** [comment, marked as answer]
I suspect the browser isn't actually connecting with http2. Hyper does not support upgrading http1 to http2, so you have to connect with http2 right away.

Try with curl's `--http2-prior-knowledge` option.

It does indeed work because [tonic](https://crates.io/crates/tonic) uses axum and http2.

**Comment by Nikaidou-Shinku** [comment]
Thank you! I may need to learn more about how http2 works, since I don't seem to understand it well.

I tried `--http2` before but failed, and I just checked the curl manual and learned that this option is also upgraded from http1.

**Comment by FrankReh** [comment]
This [answer](https://superuser.com/questions/1659248/how-does-browser-know-which-version-of-http-it-should-use-when-sending-a-request) and this [one](https://superuser.com/questions/1750003/how-does-a-web-server-communicate-http-2-support-to-a-client), both on superuser.com, seem useful in understanding how to get a browser to use HTTP/2. It is my understanding that to use HTTP/2 from a browser, one needs to be connecting to the server over https(TLS).

As part of the TLS negotiation, the browser learns the server supports HTTP/2 and generally will then try to use it first over the TLS connection.

Some browsers have a way to skip their TLS requirement for accessing localhost domains but that requires searching online for what to do for each combination of browser and OS that one is on.

**Reply by Mauricio951021** [reply]
exact
